libraries/c_sdk/standard/mqtt/src/iot_mqtt_validate.c

```diff
@@ -150,20 +153,48 @@ bool _IotMqtt_ValidateConnect( const IotMqttConnectInfo_t * pConnectInfo )
             IotLogWarn( "AWS IoT does not support disabling keep-alive. Default keep-alive "
                         "of %d seconds will be used.",
                         AWS_IOT_MQTT_SERVER_MAX_KEEPALIVE );
+
+            adjustedKeepAliveSec = AWS_IOT_MQTT_SERVER_MAX_KEEPALIVE;
         }
         else if( pConnectInfo->keepAliveSeconds < AWS_IOT_MQTT_SERVER_MIN_KEEPALIVE )
         {
             IotLogWarn( "AWS IoT does not support keep-alive intervals less than %d seconds. "
                         "An interval of %d seconds will be used.",
                         AWS_IOT_MQTT_SERVER_MIN_KEEPALIVE,
                         AWS_IOT_MQTT_SERVER_MIN_KEEPALIVE );
+
+            adjustedKeepAliveSec = AWS_IOT_MQTT_SERVER_MIN_KEEPALIVE;
         }
         else if( pConnectInfo->keepAliveSeconds > AWS_IOT_MQTT_SERVER_MAX_KEEPALIVE )
         {
             IotLogWarn( "AWS IoT does not support keep-alive intervals greater than %d seconds. "
                         "An interval of %d seconds will be used.",
                         AWS_IOT_MQTT_SERVER_MAX_KEEPALIVE,
                         AWS_IOT_MQTT_SERVER_MAX_KEEPALIVE );
+
+            adjustedKeepAliveSec = AWS_IOT_MQTT_SERVER_MAX_KEEPALIVE;
+        }
+        else
+        {
+            EMPTY_ELSE_MARKER;
+        }
+    }
+    else
+    {
+        EMPTY_ELSE_MARKER;
+    }
+
+    /* Check that keep alive is not too short. */
+    if( adjustedKeepAliveSec != 0 )
+    {
+        /* If wait time is too long, keep alive jobs will time out while waiting for PINGRESP. */
```

Comment:
Comment update:

After sending a PINGREQ, we wait for IOT_MQTT_RESPONSE_WAIT_MS to receive the corresponding PINGRESP. If the PINGRESP is received within IOT_MQTT_RESPONSE_WAIT_MS, we schedule another job to send PINGREQ. If the IOT_MQTT_RESPONSE_WAIT_MS is longer than keep alive interval, we will fail to send PINGRESP on time.

Reply:
Updated